README.md
# network-rendering
[![NPM](https://nodei.co/npm/network-rendering.png?downloads=true&stars=true)](https://www.npmjs.com/package/network-rendering)

Draw nice nodes and links in SVG. Developed for the network visualization tools at [mapequation.org](http://www.mapequation.org).

![Example image](https://cdn.rawgit.com/mapequation/network-rendering/4166608c9a292335b809faa56ecbe2dc751c29d7/example.svg)


Install
-------

```bash
npm install network-rendering
```


Example
-------

```js
import d3 from 'd3';
import networkRendering from 'network-rendering';

const linkRenderer = networkRendering.halfLink()
  .nodeRadius(node => node.size)
  .width(link => link.size)
  .oppositeLink(link => data.links[link.oppositeLink]);

svg.append("g").selectAll(".link")
    .data(data.links)
  .enter().append("path")
    .attr("class", "link")
    .style("fill", "grey")
    .style("stroke", "black")
    .attr("d", linkRenderer);
```


API
---

### networkRendering.halfLink() -> halfLinkRenderer

Creates a half-link renderer function with chainable methods.

#### halfLinkRenderer(link) -> string

Returns an SVG path string to render the `link` based on the accessor functions below.

#### halfLinkRenderer.source([sourceAccessor]) -> self

If `sourceAccessor` is specified, sets the source node accessor to the specified function. If not specified, returns the current source node accessor, which defaults to:

```js
(link) => link.source
```

#### halfLinkRenderer.target([targetAccessor]) -> self

If `targetAccessor` is specified, sets the target node accessor to the specified function. If not specified, returns the current target node accessor, which defaults to:

```js
(link) => link.target
```

#### halfLinkRenderer.nodeRadius([radius]) -> self

If `radius` is specified, sets the node radius accessor to the specified function or constant. If not specified, returns the current node radius accessor, which defaults to:

```js
(node) => node.size || 10
```

#### halfLinkRenderer.nodeX([x]) -> self

If `x` is specified, sets the nodes' x-coordinate accessor to the specified function or constant. If not specified, returns the nodes' current x-coordinate accessor, which defaults to:

```js
(node) => node.x
```

#### halfLinkRenderer.nodeY([y]) -> self

If `y` is specified, sets the nodes' y-coordinate accessor to the specified function or constant. If not specified, returns the nodes' current y-coordinate accessor, which defaults to:

```js
(node) => node.y
```

#### halfLinkRenderer.width([width]) -> self

If `width` is specified, sets the width accessor to the specified function or constant. If not specified, returns the current width accessor, which defaults to:

```js
(link) => link.size || 10
```

#### halfLinkRenderer.oppositeLink([oppositeLink]) -> self

If `oppositeLink` is specified, sets the opposite link accessor to the specified function. If not specified, returns the current opposite link accessor, which defaults to:

```js
(link) => null
```

If the link renderer can't access opposite links, existing opposite links may not be rendered to fit nicely together.


#### halfLinkRenderer.bend([bend]) -> self

If `bend` is specified, sets the bend accessor to the specified function or constant. If not specified, returns the current bend accessor, which defaults to:

```js
(link) => link.bend || 30
```

### networkRendering.undirectedLink() -> undirectedLinkRenderer

Creates a undirected-link renderer function with chainable methods.
The API is the same as for halfLinkRenderer except that oppositeLink does not exist in undirectedLinkRenderer.


License
-------

[MPL-2.0](https://github.com/mapequation/network-rendering/blob/master/LICENSE)


CHANGELOG.md
### [1.0.0](https://github.com/mapequation/network-rendering/releases/tag/v1.0.0)

- Support undirected and directed links with bending and merging of opposite directed links


test/unit/svgPath.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import SvgPath, * as primitives from '../../src/utils/SvgPath';

describe('SvgPath', () => {
  describe('svg path helpers', () => {
    it('should return correct primitive svg path strings', () => {
      expect(primitives.M(1,2)).to.equal('M 1 2');
      expect(primitives.m(1,2)).to.equal('m 1 2');
      expect(primitives.L(1,2)).to.equal('L 1 2');
      expect(primitives.l(1,2)).to.equal('l 1 2');
      expect(primitives.Q(1,2,3,4)).to.equal('Q 1 2, 3 4');
      expect(primitives.Z()).to.equal('Z');
    });
    it('should return correct commplex svg path strings', () => {
      expect(new SvgPath()
        .M(1,2)
        .L(3,4)
        .Q(5,6,7,8)
        .Z()
        .getString()
      ).to.equal('M 1 2 L 3 4 Q 5 6, 7 8 Z');
    });
    it('should return default absolute svg path', () => {
      expect(new SvgPath()
        .moveTo(1,2)
        .lineTo(3,4)
        .curveTo(5,6,7,8)
        .end()
        .getString()
      ).to.equal('M 1 2 L 3 4 Q 5 6, 7 8 Z');
    });
    it('should return correct absolute svg path', () => {
      expect(new SvgPath()
        .absolute()
        .moveTo(1,2)
        .lineTo(3,4)
        .curveTo(5,6,7,8)
        .end()
        .getString()
      ).to.equal('M 1 2 L 3 4 Q 5 6, 7 8 Z');
    });
    it('should return correct relative svg path', () => {
      expect(new SvgPath()
        .relative()
        .moveTo(1,2)
        .lineTo(3,4)
        .curveTo(5,6,7,8)
        .end()
        .getString()
      ).to.equal('m 1 2 l 3 4 q 5 6, 7 8 Z');
    });
    it('should return correct mix of absolute and relative svg path', () => {
      expect(new SvgPath()
        .moveTo(1,2)
        .relative()
        .lineTo(3,4)
        .lineTo(5,6)
        .absolute()
        .lineTo(7,8)
        .lineTo(9,10)
        .end()
        .getString()
      ).to.equal('M 1 2 l 3 4 l 5 6 L 7 8 L 9 10 Z');
    });
  });
});


test/unit/network-rendering.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import networkRendering from '../../src/network-rendering';

function checkLinkPath(linkPath) {
  expect(linkPath).to.be.a('string');
  expect(linkPath.length).to.be.within(100, 500);
  expect(linkPath.charAt(0)).to.equal('M');
  expect(linkPath.charAt(linkPath.length - 1)).to.equal('Z');
  expect(linkPath).to.not.contain('NaN');
}

describe('networkRendering', () => {
  describe('halfLink', () => {
    var linkRenderer;

    beforeEach(() => {
      spy(networkRendering, 'halfLink');
      linkRenderer = networkRendering.halfLink();
      spy(linkRenderer, 'width');
      spy(linkRenderer, 'bend');
    });

    it('halfLink should have been called once', () => {
      expect(networkRendering.halfLink).to.have.been.calledOnce;
    });

    it('should return empty link for overlapping nodes', () => {
      const linkPath = linkRenderer({
        source: {x: 10, y: 10, size: 10},
        target: {x: 20, y: 20, size: 10}
      });
      expect(linkPath).to.equal('');
    });

    it('should return non-empty link for overlapping nodes if large bend', () => {
      linkRenderer.bend(50);
      const linkPath = linkRenderer({
        source: {x: 10, y: 10, size: 10},
        target: {x: 20, y: 20, size: 10}
      });
      expect(linkRenderer.bend).to.have.been.calledOnce;
      expect(linkRenderer.bend()('dummy link')).to.equal(50);
      checkLinkPath(linkPath);
      linkRenderer.bend(20);
      expect(linkRenderer.bend).to.have.been.calledThrice;
    });

    it('should return a SVG path string for non-overlapping nodes', () => {
      const linkPath = linkRenderer({
        source: {x: 10, y: 10, size: 10},
        target: {x: 30, y: 20, size: 5}
      });
      expect(linkPath).to.contain('M');
      expect(linkPath.length).to.be.within(100, 500);
    });
  });

  describe('undirectedLink', () => {
    var linkRenderer;

    beforeEach(() => {
      spy(networkRendering, 'undirectedLink');
      linkRenderer = networkRendering.undirectedLink();
      spy(linkRenderer, 'width');
      spy(linkRenderer, 'bend');
    });

    it('undirectedLink should have been called once', () => {
      expect(networkRendering.undirectedLink).to.have.been.calledOnce;
    });

    it('should return empty link for overlapping nodes', () => {
      const linkPath = linkRenderer({
        source: {x: 10, y: 10, size: 10},
        target: {x: 20, y: 20, size: 10}
      });
      expect(linkPath).to.equal('');
    });

    it('should return non-empty link for overlapping nodes if large bend', () => {
      linkRenderer.bend(50);
      const linkPath = linkRenderer({
        source: {x: 10, y: 10, size: 10},
        target: {x: 20, y: 20, size: 10}
      });
      expect(linkRenderer.bend).to.have.been.calledOnce;
      expect(linkRenderer.bend()('dummy link')).to.equal(50);
      checkLinkPath(linkPath);
      linkRenderer.bend(20);
      expect(linkRenderer.bend).to.have.been.calledThrice;
    });

    it('should return a SVG path string for non-overlapping nodes', () => {
      const linkPath = linkRenderer({
        source: {x: 10, y: 10, size: 10},
        target: {x: 30, y: 20, size: 5}
      });
      expect(linkPath).to.contain('M');
      expect(linkPath.length).to.be.within(100, 500);
    });
  });
});


gulpfile.babel.js
import gulp  from 'gulp';
import loadPlugins from 'gulp-load-plugins';
import del  from 'del';
import glob  from 'glob';
import path  from 'path';
import {Instrumenter} from 'isparta';
import webpack from 'webpack';
import webpackStream from 'webpack-stream';
import source  from 'vinyl-source-stream';

import mochaGlobals from './test/setup/.globals';
import manifest  from './package.json';

// Load all of our Gulp plugins
const $ = loadPlugins();

// Gather the library data from `package.json`
const config = manifest.babelBoilerplateOptions;
const mainFile = manifest.main;
const destinationFolder = path.dirname(mainFile);
const exportFileName = path.basename(mainFile, path.extname(mainFile));

function cleanDist(done) {
  del([destinationFolder]).then(() => done());
}

function cleanTmp(done) {
  del(['tmp']).then(() => done());
}

function onError(err) {
  $.util.beep();
  // console.log('!!!! ERROR !!!:', err);
  $.notify().write(err);
}

// Lint a set of files
function lint(files) {
  return gulp.src(files)
    .pipe($.plumber())
    .pipe($.eslint())
    .pipe($.eslint.format())
    .pipe($.eslint.failOnError())
    .pipe($.jscs())
    .pipe($.jscs.reporter())
    .pipe($.jscs.reporter('fail'))
    .on('error', onError);
}

function lintSrc() {
  return lint('src/**/*.js');
}

function lintTest() {
  return lint('test/**/*.js');
}

function lintGulpfile() {
  return lint('gulpfile.babel.js');
}

function build() {
  return gulp.src(path.join('src', config.entryFileName + '.js'))
    .pipe($.plumber())
    .pipe(webpackStream({
      output: {
        filename: exportFileName + '.js',
        libraryTarget: 'umd',
        library: config.mainVarName
      },
      module: {
        loaders: [
          { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' }
        ]
      },
      devtool: 'source-map'
    }))
    .pipe(gulp.dest(destinationFolder))
    .pipe($.filter(['*', '!**/*.js.map']))
    .pipe($.rename(exportFileName + '.min.js'))
    .pipe($.sourcemaps.init({ loadMaps: true }))
    .pipe($.uglify())
    .pipe($.sourcemaps.write('./'))
    .pipe(gulp.dest(destinationFolder));
}

function _mocha() {
  return gulp.src(['test/setup/node.js', 'test/unit/**/*.js'], {read: false})
    .pipe($.mocha({
      reporter: 'dot',
      globals: Object.keys(mochaGlobals.globals),
      ignoreLeaks: false
    }))
    .on('error', function(err) {
      $.notify().write(err);
    });
}

function _registerBabel() {
  require('babel-register');
}

function test() {
  _registerBabel();
  return _mocha();
}

function coverage(done) {
  _registerBabel();
  gulp.src(['src/**/*.js'])
    .pipe($.istanbul({ instrumenter: Instrumenter }))
    .pipe($.istanbul.hookRequire())
    .on('finish', () => {
      return test()
        .pipe($.istanbul.writeReports())
        .on('end', done);
    });
}

const watchFiles = ['src/**/*', 'test/**/*', 'package.json', '**/.eslintrc', '.jscsrc'];

// Run the headless unit tests as you make changes.
function watch() {
  gulp.watch(watchFiles, ['test']);
}

function watchBuild() {
  gulp.watch(watchFiles, ['test', 'build']);
}

function testBrowser() {
  // Our testing bundle is made up of our unit tests, which
  // should individually load up pieces of our application.
  // We also include the browser setup file.
  const testFiles = glob.sync('./test/unit/**/*.js');
  const allFiles = ['./test/setup/browser.js'].concat(testFiles);

  // Lets us differentiate between the first build and subsequent builds
  var firstBuild = true;

  // This empty stream might seem like a hack, but we need to specify all of our files through
  // the `entry` option of webpack. Otherwise, it ignores whatever file(s) are placed in here.
  return gulp.src('')
    .pipe($.plumber())
    .pipe(webpackStream({
      watch: true,
      entry: allFiles,
      output: {
        filename: '__spec-build.js'
      },
      module: {
        loaders: [
          // This is what allows us to author in future JavaScript
          { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' },
          // This allows the test setup scripts to load `package.json`
          { test: /\.json$/, exclude: /node_modules/, loader: 'json-loader' }
        ]
      },
      plugins: [
        // By default, webpack does `n=>n` compilation with entry files. This concatenates
        // them into a single chunk.
        new webpack.optimize.LimitChunkCountPlugin({ maxChunks: 1 })
      ],
      devtool: 'inline-source-map'
    }, null, function() {
      if (firstBuild) {
        $.livereload.listen({port: 35729, host: 'localhost', start: true});
        var watcher = gulp.watch(watchFiles, ['lint']);
      } else {
        $.livereload.reload('./tmp/__spec-build.js');
      }
      firstBuild = false;
    }))
    .pipe(gulp.dest('./tmp'));
}

// Remove the built files
gulp.task('clean', cleanDist);

// Remove our temporary files
gulp.task('clean-tmp', cleanTmp);

// Lint our source code
gulp.task('lint-src', lintSrc);

// Lint our test code
gulp.task('lint-test', lintTest);

// Lint this file
gulp.task('lint-gulpfile', lintGulpfile);

// Lint everything
gulp.task('lint', ['lint-src', 'lint-test', 'lint-gulpfile']);

// Build two versions of the library
gulp.task('build', ['lint', 'clean'], build);

// Lint and run our tests
gulp.task('test', ['lint'], test);

// Set up coverage and run tests
gulp.task('coverage', ['lint'], coverage);

// Set up a livereload environment for our spec runner `test/runner.html`
gulp.task('test-browser', ['lint', 'clean-tmp'], testBrowser);

// Run the headless unit tests as you make changes.
gulp.task('watch', watch);

// Run the headless unit tests as you make changes.
gulp.task('watch-build', watchBuild);

// An alias of test
gulp.task('default', ['test']);


src/utils/svgPath.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

export const M = (x, y) => `M ${x} ${y}`;
export const m = (x, y) => `m ${x} ${y}`;
export const L = (x, y) => `L ${x} ${y}`;
export const l = (x, y) => `l ${x} ${y}`;
export const Q = (x1, y1, x, y) => `Q ${x1} ${y1}, ${x} ${y}`;
export const q = (x1, y1, x, y) => `q ${x1} ${y1}, ${x} ${y}`;
export const Z = () => `Z`;

export default class SvgPath {
  constructor() {
    this._path = [];
    this._relative = false;
  }

  getString() { return this._path.join(' '); }

  relative() { this._relative = true; return this; }
  absolute() { this._relative = false; return this; }

  m(x, y) { this._path.push(m(x,y)); return this; }
  M(x, y) { this._path.push(M(x,y)); return this; }
  moveTo(x, y) { this._path.push(this._relative ? m(x, y) : M(x, y)); return this; }

  l(x, y) { this._path.push(l(x, y)); return this; }
  L(x, y) { this._path.push(L(x, y)); return this; }
  lineTo(x, y) { this._path.push(this._relative ? l(x, y) : L(x, y)); return this; }

  q(x1, y1, x, y) { this._path.push(q(x1, y1, x, y)); return this; }
  Q(x1, y1, x, y) { this._path.push(Q(x1, y1, x, y)); return this; }
  curveTo(x1, y1, x, y) { this._path.push(this._relative ? q(x1, y1, x, y) : Q(x1, y1, x, y)); return this; }

  Z() { this._path.push(Z()); return this; }
  end() { this._path.push(Z()); return this; }
}


src/network-rendering.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import SvgPath from './utils/SvgPath';
import functor from './utils/functor';

export function halfLink() {

  var _source = (link) => link.source;
  var _target = (link) => link.target;
  var _nodeRadius = (node) => node.size || 10;
  var _nodeX = (node) => node.x;
  var _nodeY = (node) => node.y;
  var _width = (link) => link.width || 10;
  var _bend = (link) => link.bend || 30;
  var _oppositeLink = (link) => null;

  renderHalfLink.source = function(_) { if (!arguments.length) { return _source; } _source = functor(_); return renderHalfLink; };
  renderHalfLink.target = function(_) { if (!arguments.length) { return _target; } _target = functor(_); return renderHalfLink; };
  renderHalfLink.nodeRadius = function(_) { if (!arguments.length) { return _nodeRadius; } _nodeRadius = functor(_); return renderHalfLink; };
  renderHalfLink.nodeX = function(_) { if (!arguments.length) { return _nodeX; } _nodeX = functor(_); return renderHalfLink; };
  renderHalfLink.nodeY = function(_) { if (!arguments.length) { return _nodeY; } _nodeY = functor(_); return renderHalfLink; };
  renderHalfLink.width = function(_) { if (!arguments.length) { return _width; } _width = functor(_); return renderHalfLink; };
  renderHalfLink.bend = function(_) { if (!arguments.length) { return _bend; } _bend = functor(_); return renderHalfLink; };
  renderHalfLink.oppositeLink = function(_) { if (!arguments.length) { return _oppositeLink; } _oppositeLink = functor(_); return renderHalfLink; };

  function renderHalfLink(link) {
    const source = _source(link);
    const target = _target(link);
    const x0 = _nodeX(source);
    const y0 = _nodeY(source);
    const r0 = _nodeRadius(source);
    const x1 = _nodeX(target);
    const y1 = _nodeY(target);
    const r1 = _nodeRadius(target);
    const width = _width(link);
    const oppositeLink = _oppositeLink(link);
    const bend = _bend(link);
    const oppositeWidth = oppositeLink ? _width(oppositeLink) : width;
    const dx = x1 - x0;
    const dy = y1 - y0;
    const l = Math.sqrt(dx * dx + dy * dy);
    const lBetween = l - r0 - r1;

    // Skip draw link of nodes overlap if not big bend
    if (lBetween <= 0 && Math.abs(bend) < 50) {
      return '';
    }

    // Get unit vector in direction and perpendicular
    const dir = {x: dx / l, y: dy / l};
    const right = {x: -dir.y, y: dir.x};

    // Size of arrowhead
    const tipLength = Math.min(lBetween / 3, 10 * Math.pow(width, 1 / 3));
    const tipWidth = 2 * Math.pow(width, 1 / 2); // excluding the line width

    // Size of opposite arrowhead
    const oppositeTipLength = Math.min(lBetween / 3, 10 * Math.pow(oppositeWidth, 1 / 3));

    // Bending
    const bendMagnitude = Math.abs(bend);
    const outerBendAddition = Math.pow(bendMagnitude / 10, 0.4);
    const positiveCurvature = dir.x > 0 || (dir.x == 0 && dir.y < 0);
    const curvatureSign = positiveCurvature ? 1 : -1;
    const bendSign = bend > 0 ? 1 : -1;
    const finalBendSign = curvatureSign * bendSign;
    const signedBend = finalBendSign * bendMagnitude;

    // Calculate the end points for the middle bezier curve. Rotate them towards the control point later.
    const x02tmp = x0 + (r0 + oppositeTipLength) * dir.x;
    const y02tmp = y0 + (r0 + oppositeTipLength) * dir.y;
    const x12tmp = x1 - (r1 + tipLength) * dir.x;
    const y12tmp = y1 - (r1 + tipLength) * dir.y;

    const xMidpoint = 0.5 * (x02tmp + x12tmp);
    const yMidpoint = 0.5 * (y02tmp + y12tmp);
    const xCP1 = xMidpoint + signedBend * right.x;
    const yCP1 = yMidpoint + signedBend * right.y;
    const xCP2 = xMidpoint + (signedBend + width + outerBendAddition) * right.x;
    const yCP2 = yMidpoint + (signedBend + width + outerBendAddition) * right.y;

    // points from source to control point
    const dx1 = xCP1 - x0;
    const dy1 = yCP1 - y0;
    const l1 = Math.sqrt(dx1 * dx1 + dy1 * dy1);
    const dir0 = {x: dx1 / l1, y: dy1 / l1};
    const right0 = {x: -dir0.y, y: dir0.x};
    const x01 = x0 + r0 * dir0.x;
    const y01 = y0 + r0 * dir0.y;
    const x02 = x01 + oppositeTipLength * dir0.x;
    const y02 = y01 + oppositeTipLength * dir0.y;
    const x03 = x02 + width * right0.x;
    const y03 = y02 + width * right0.y;
    const x04 = x0 + width * right0.x;
    const y04 = y0 + width * right0.y;

    // points from target to control point
    const dx2 = xCP1 - x1;
    const dy2 = yCP1 - y1;
    const l2 = Math.sqrt(dx2 * dx2 + dy2 * dy2);
    const dir1 = {x: dx2 / l2, y: dy2 / l2};
    const x11 = x1 + r1 * dir1.x;
    const y11 = y1 + r1 * dir1.y;
    const x12 = x11 + tipLength * dir1.x;
    const y12 = y11 + tipLength * dir1.y;
    const left1 = {x: dir1.y, y: -dir1.x};
    const x13 = x12 + width * left1.x;
    const y13 = y12 + width * left1.y;
    const x14 = x13 + tipWidth * left1.x;
    const y14 = y13 + tipWidth * left1.y;

    return new SvgPath()
      .moveTo(x02, y02)
      .lineTo(x0, y0)
      .lineTo(x04, y04)
      .lineTo(x03, y03)
      .curveTo(xCP2, yCP2, x13, y13)
      .lineTo(x14, y14)
      .lineTo(x11, y11)
      .lineTo(x12, y12)
      .curveTo(xCP1, yCP1, x02, y02)
      .end()
      .getString();
  }

  return renderHalfLink;
}

export function undirectedLink() {
  let _source = (link) => link.source;
  let _target = (link) => link.target;
  let _nodeRadius = (node) => node.size || 10;
  let _nodeX = (node) => node.x;
  let _nodeY = (node) => node.y;
  let _width = (link) => link.width || 10;
  let _bend = (link) => link.bend || 30;

  renderUndirectedLink.source = function(_) { if (!arguments.length) { return _source; } _source = functor(_); return renderUndirectedLink; };
  renderUndirectedLink.target = function(_) { if (!arguments.length) { return _target; } _target = functor(_); return renderUndirectedLink; };
  renderUndirectedLink.nodeRadius = function(_) { if (!arguments.length) { return _nodeRadius; } _nodeRadius = functor(_); return renderUndirectedLink; };
  renderUndirectedLink.nodeX = function(_) { if (!arguments.length) { return _nodeX; } _nodeX = functor(_); return renderUndirectedLink; };
  renderUndirectedLink.nodeY = function(_) { if (!arguments.length) { return _nodeY; } _nodeY = functor(_); return renderUndirectedLink; };
  renderUndirectedLink.width = function(_) { if (!arguments.length) { return _width; } _width = functor(_); return renderUndirectedLink; };
  renderUndirectedLink.bend = function(_) { if (!arguments.length) { return _bend; } _bend = functor(_); return renderUndirectedLink; };

  function renderUndirectedLink(link) {
    const source = _source(link);
    const target = _target(link);
    const x0 = _nodeX(source);
    const y0 = _nodeY(source);
    const r0 = _nodeRadius(source);
    const x1 = _nodeX(target);
    const y1 = _nodeY(target);
    const r1 = _nodeRadius(target);
    const width = _width(link);
    const bend = _bend(link);
    const dx = x1 - x0;
    const dy = y1 - y0;
    const l = Math.sqrt(dx * dx + dy * dy);
    const lBetween = l - r0 - r1;

    // Skip draw link of nodes overlap if not big bend
    if (lBetween <= 0 && Math.abs(bend) < 50) {
      return '';
    }

    // Get unit vector in direction and perpendicular
    const dir = { x: dx / l, y: dy / l };
    const right = { x: -dir.y, y: dir.x };

    // Bending
    const bendMagnitude = Math.abs(bend);
    const outerBendAddition = Math.pow(bendMagnitude / 10,  0.4);
    const positiveCurvature = dir.x > 0 || (dir.x === 0 && dir.y < 0);
    const curvatureSign = positiveCurvature ? 1 : -1;
    const bendSign = bend > 0 ? 1 : -1;
    const finalBendSign = curvatureSign * bendSign;
    const signedBend = finalBendSign * bendMagnitude;

    // Calculate the end points for the middle bezier curve. Rotate them towards the control point later.
    const x02tmp = x0 + r0 * dir.x;
    const y02tmp = y0 + r0 * dir.y;
    const x12tmp = x1 - r1 * dir.x;
    const y12tmp = y1 - r1 * dir.y;

    const xMidpoint = 0.5 * (x02tmp + x12tmp);
    const yMidpoint = 0.5 * (y02tmp + y12tmp);
    const xCP1 = xMidpoint + signedBend * right.x;
    const yCP1 = yMidpoint + signedBend * right.y;
    const xCP2 = xMidpoint + (signedBend + width + outerBendAddition) * right.x;
    const yCP2 = yMidpoint + (signedBend + width + outerBendAddition) * right.y;

    const halfWidth = 0.5 * width;

    // points from source to control point
    const dx1 = xCP1 - x0;
    const dy1 = yCP1 - y0;
    const l1 = Math.sqrt(dx1 * dx1 + dy1 * dy1);
    const dir0 = { x: dx1 / l1, y: dy1 / l1 };
    const right0 = { x: -dir0.y, y: dir0.x };
    const x01 = x0 - halfWidth * right0.x;
    const y01 = y0 - halfWidth * right0.y;
    const x02 = x0 + halfWidth * right0.x;
    const y02 = y0 + halfWidth * right0.y;

    // points from target to control point
    const dx2 = xCP1 - x1;
    const dy2 = yCP1 - y1;
    const l2 = Math.sqrt(dx2 * dx2 + dy2 * dy2);
    const dir1 = { x: dx2 / l2, y: dy2 / l2 };
    const left1 = { x: dir1.y, y: -dir1.x };
    const x11 = x1 - halfWidth * left1.x;
    const y11 = y1 - halfWidth * left1.y;
    const x12 = x1 + halfWidth * left1.x;
    const y12 = y1 + halfWidth * left1.y;

    return new SvgPath()
      .moveTo(x01, y01)
      .lineTo(x02, y02)
      .curveTo(xCP2, yCP2, x12, y12)
      .lineTo(x11, y11)
      .curveTo(xCP1, yCP1, x01, y01)
      .end()
      .getString();
  }

  return renderUndirectedLink;
}

const networkRendering = {
  halfLink,
  undirectedLink,
};

export default networkRendering;



example/network.js

var data = {
  nodes: [
    {
      x: 100,
      y: 100,
      flow: 0.6,
      outFlow: 0.3
    },
    {
      x: 300,
      y: 180,
      flow: 0.4,
      outFlow: 0.2
    }
  ],
  links: [
    {
      source: 0,
      target: 1,
      flow: 0.5
    },
    {
      source: 1,
      target: 0,
      flow: 0.3
    }
  ]
};

// Connect network
data.links.forEach(function(link) {
  link.source = data.nodes[link.source];
  link.target = data.nodes[link.target];
});

// Connect opposite links
data.links[0].oppositeLink = 1;
data.links[1].oppositeLink = 0;

var width = 400,
    height = 300;

var nodeFillColor = d3.scale.linear()
  .domain(d3.extent(data.nodes, function(d) { return d.flow; }))
  .range(["#EF7518", "#D75908"]);
var nodeBorderColor = d3.scale.linear()
  .domain(d3.extent(data.nodes, function(d) { return d.outFlow; }))
  .range(["#FFAE38", "#f9a327"]);
var nodeRadius = d3.scale.linear()
  .domain(d3.extent(data.nodes, function(d) { return d.flow; }))
  .range([20, 30]);
var nodeBorderWidth = d3.scale.linear()
  .domain(d3.extent(data.nodes, function(d) { return d.outFlow; }))
  .range([3, 6]);
var linkFillColor = d3.scale.linear()
  .domain(d3.extent(data.links, function(d) { return d.flow; }))
  .range(["#71B2D7", "#418EC7"]);
var linkWidth = d3.scale.linear()
  .domain(d3.extent(data.links, function(d) { return d.flow; }))
  .range([7, 13]);

var linkRenderer = networkRendering.halfLink()
  .nodeRadius(function(node) { return nodeRadius(node.flow); })
  .width(function(d) { return linkWidth(d.flow); })
  .oppositeLink(function(d) { return data.links[d.oppositeLink]; });

var svg = d3.select("body").append("svg")
    .attr("width", width)
    .attr("height", height);

var links = svg.append("g").selectAll(".link")
    .data(data.links)
  .enter().append("path")
    .attr("class", "link")
    .style("fill", function(d) { return linkFillColor(d.flow); })
    .style("stroke", "none")
    .style("stroke-width", "1.5px")
    .attr("d", linkRenderer);

var nodes = svg.append("g").selectAll(".node")
    .data(data.nodes)
  .enter().append("circle")
    .attr("class", "node")
    .style("fill", function(d) { return nodeFillColor(d.flow); })
    .style("stroke", function(d) { return nodeBorderColor(d.outFlow); })
    .style("stroke-width", function(d) { return nodeBorderWidth(d.outFlow); })
    .attr("cx", function(d) { return d.x; })
    .attr("cy", function(d) { return d.y; })
    .attr("r", function(d) { return nodeRadius(d.flow); });
